sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Helpers/SonarAnalysisContext.cs

```diff
@@ -145,57 +152,57 @@ internal void RegisterCodeBlockStartAction<TLanguageKindEnum>(Action<CodeBlockSt
+                        if (!IsAnalysisDisabled(c.GetSyntaxTree()))
+                        {
+                            action(c);
+                        }
+                    }, syntaxKinds));
 
         internal void RegisterSyntaxTreeAction(Action<SyntaxTreeAnalysisContext> action) =>
-            context.RegisterSyntaxTreeAction(
-                c =>
-                {
-                    if (!IsAnalysisDisabled(c.GetSyntaxTree(), supportedDiagnostics))
+            RegisterContextAction(() =>
+                context.RegisterSyntaxTreeAction(
+                    c =>
                     {
-                        action(c);
-                    }
-                });
+                        if (!IsAnalysisDisabled(c.GetSyntaxTree()))
+                        {
+                            action(c);
+                        }
+                    }));
 
         internal void RegisterSymbolAction(Action<SymbolAnalysisContext> action, ImmutableArray<SymbolKind> symbolKinds) =>
             RegisterSymbolAction(action, symbolKinds.ToArray());
 
         public void RegisterSymbolAction(Action<SymbolAnalysisContext> action, params SymbolKind[] symbolKinds) =>
-            context.RegisterSymbolAction(
-                c =>
-                {
-                    if (!IsAnalysisDisabled(c.GetSyntaxTree(), supportedDiagnostics))
+            RegisterContextAction(() =>
+                context.RegisterSymbolAction(
+                    c =>
                     {
-                        action(c);
-                    }
-                }, symbolKinds);
+                        if (!IsAnalysisDisabled(c.GetSyntaxTree()))
+                        {
+                            action(c);
+                        }
+                    }, symbolKinds));
 
-        internal static bool IsAnalysisDisabled(SyntaxTree tree, IEnumerable<DiagnosticDescriptor> supportedDiagnostics = null)
+        private void RegisterContextAction(Action action)
         {
-            // This is the new way SonarLint will handle whether or not the analysis should be performed...
-            // (checking `supportedDiagnostics != null` is to force providing the old behavior for SonarCodeFixProvider)
-            if (ShouldExecuteRuleFunc != null &&
-                supportedDiagnostics != null)
+            if (ShouldRegisterContextAction?.Invoke(this.supportedDiagnostics) != false)
             {
-                Debug.Assert(ShouldAnalysisBeDisabled == null, "Not expecting SonarLint to set both the old and the new" +
-                    "delegates.");
+                action();
+            }
+        }
 
-                // If any of the diagnostic is enabled then allow to run the rule action BUT filter out at the time when the
-                // issue is being reported.
-                return !ShouldExecuteRuleFunc(new AnalysisRunContext(tree, supportedDiagnostics));
+        internal static bool IsAnalysisDisabled(SyntaxTree tree)

```

Comment:
I only reverted the change but I could probably change the method to be positive.